MREUnityRuntime/MREUnityRuntimeLib/Behaviors/Contexts/ButtonBehaviorContext.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using MixedRealityExtension.Behaviors.ActionData;
+using MixedRealityExtension.Behaviors.Actions;
+using MixedRealityExtension.Core.Interfaces;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace MixedRealityExtension.Behaviors.Contexts
+{
+	public class ButtonBehaviorContext : TargetBehaviorContext
+	{
+		private MWAction<ButtonData> _hoverAction = new MWAction<ButtonData>();
+		private MWAction<ButtonData> _clickAction = new MWAction<ButtonData>();
+		private MWAction<ButtonData> _buttonAction = new MWAction<ButtonData>();
+		private List<Vector3> _buttonPressedPoints = new List<Vector3>();
+		private List<Vector3> _hoverPoints = new List<Vector3>();
+
+		public bool IsPressed { get; private set; } 
+
+		public void StartHover(IUser user, Vector3 hoverPoint)
+		{
+			var app = App;
+			if (app == null)
+			{
+				return;
+			}
+
+			_hoverAction.StartAction(user, new ButtonData()
+			{
+				targetedPoints = new PointData[1]
+				{
+					PointData.CreateFromUnityVector3(hoverPoint, Behavior.Actor.GameObject.transform, App.SceneRoot.transform)
+				}
+			});
+		}
+
+		public void EndHover(IUser user, Vector3 hoverPoint)
+		{
+			var app = App;
+			if (app == null)
+			{
+				return;
+			}
+
+			_hoverAction.StopAction(user, new ButtonData()
+			{
+				targetedPoints = new PointData[1]
+				{
+					PointData.CreateFromUnityVector3(hoverPoint, Behavior.Actor.GameObject.transform, App.SceneRoot.transform)
+				}
+			});
+		}
+
+		public void StartButton(IUser user, Vector3 buttonStartPoint)
+		{
+			var app = App;
+			if (app == null)
+			{
+				return;
+			}
+
+			_buttonAction.StartAction(user, new ButtonData()
+			{
+				targetedPoints = new PointData[1]
+				{
+					PointData.CreateFromUnityVector3(buttonStartPoint, Behavior.Actor.GameObject.transform, App.SceneRoot.transform)
+				}
+			});
+
+			IsPressed = true;
+		}
+
+		public void EndButton(IUser user, Vector3 buttonEndPoint)
+		{
+			var app = App;
+			if (app == null)
+			{
+				return;
+			}
+
+			_buttonAction.StopAction(user, new ButtonData()
+			{
+				targetedPoints = new PointData[1]
+				{
+					PointData.CreateFromUnityVector3(buttonEndPoint, Behavior.Actor.GameObject.transform, App.SceneRoot.transform)
+				}
+			});
+
+			IsPressed = false;
+		}
+
+		public void Click(IUser user, Vector3 clickPoint)
+		{
+			var app = App;
+			if (app == null)
+			{
+				return;
+			}
+
+			_clickAction.StartAction(user, new ButtonData()
+			{
+				targetedPoints = new PointData[1]
+				{
+					PointData.CreateFromUnityVector3(clickPoint, Behavior.Actor.GameObject.transform, App.SceneRoot.transform)
+				}
+			});
+		}
+
+		internal ButtonBehaviorContext()
```

Comment:
Empty constructor is unnecessary, and base() is called automatically.

Reply:
There used to be things done in the constructor.  Now there are not.  Will clean up.

---
In reply to: [428857200](https://github.com/microsoft/mixed-reality-extension-unity/pull/160#discussion_r428857200) [](ancestors = 428857200)

Reply:
Actually now I remember why I had added constructors.  It was to scope them to internal accessibility so not to expose a way to allocate out side of the runtime library code.  The class is public and accessible to the host app. 

---
In reply to: [434091182](https://github.com/microsoft/mixed-reality-extension-unity/pull/160#discussion_r434091182) [](ancestors = 434091182,428857200)